File: launcher/ffmpeg_config.go
// FFmpeg config file

package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path"
	"runtime"
	"strings"

	"github.com/nicksnyder/go-i18n/v2/i18n"
)

type FFmpegConfig struct {
	FFProbePath string `json:"ffmpeg_path"`
	FFMpegPath  string `json:"ffprobe_path"`
	H264Codec   string `json:"h264_codec"`
}

func loadFFmpegConfigFromFile() (*FFmpegConfig, error) {
	userConfigDir, err := os.UserConfigDir()

	if err != nil {
		return nil, err
	}

	vaultConfigDir := path.Join(userConfigDir, "PersonalMediaVault")

	err = os.MkdirAll(vaultConfigDir, FOLDER_PERMISSION)

	if err != nil {
		return nil, err
	}

	file := path.Join(vaultConfigDir, "ffmpeg.json")

	b, err := os.ReadFile(file)

	if err != nil {
		return nil, err
	}

	config := FFmpegConfig{}

	// Parse
	err = json.Unmarshal(b, &config)

	if err != nil {
		return nil, err
	}

	return &config, nil
}

func writeFFmpegConfigToFile(config *FFmpegConfig) error {
	userConfigDir, err := os.UserConfigDir()

	if err != nil {
		return err
	}

	vaultConfigDir := path.Join(userConfigDir, "PersonalMediaVault")

	err = os.MkdirAll(vaultConfigDir, FOLDER_PERMISSION)

	if err != nil {
		return err
	}

	file := path.Join(vaultConfigDir, "ffmpeg.json")

	b, err := json.Marshal(config)

	if err != nil {
		return err
	}

	return os.WriteFile(file, b, FILE_PERMISSION)
}

const H264_CODEC_DEFAULT = "libx264"
const H264_CODEC_FREE = "libopenh264"

func detectH264Codec(config *FFmpegConfig) (string, error) {
	out, err := exec.Command(config.FFMpegPath, "-encoders").Output()

	if err != nil {
		return "", err
	}

	availableCodecs := make(map[string]bool)

	lines := strings.Split(string(out), "\n")

	for _, line := range lines {
		lineParts := strings.Split(strings.TrimSpace(line), " ")

		if len(lineParts) < 2 {
			continue
		}

		codecName := lineParts[1]

		availableCodecs[codecName] = true
	}

	if availableCodecs[H264_CODEC_DEFAULT] {
		return H264_CODEC_DEFAULT, nil
	}

	if availableCodecs[H264_CODEC_FREE] {
		return H264_CODEC_FREE, nil
	}

	return "", errors.New("unavailable codec")
}

func checkFFmpegCodec(config *FFmpegConfig) bool {
	out, err := exec.Command(config.FFMpegPath, "-encoders").Output()

	if err != nil {
		return false
	}

	availableCodecs := make(map[string]bool)

	lines := strings.Split(string(out), "\n")

	for _, line := range lines {
		lineParts := strings.Split(strings.TrimSpace(line), " ")

		if len(lineParts) < 2 {
			continue
		}

		codecName := lineParts[1]

		availableCodecs[codecName] = true
	}

	return availableCodecs[config.H264Codec]
}

func loadFFmpegConfig() *FFmpegConfig {
	result := &FFmpegConfig{
		FFProbePath: "",
		FFMpegPath:  "",
		H264Codec:   "",
	}

	configFromFile, err := loadFFmpegConfigFromFile()

	if err == nil {
		result = configFromFile
	}

	// Check

	if !fileExists(result.FFMpegPath) {
		result.FFMpegPath = path.Join(getDirName(), "bin", getBinaryFileName("ffmpeg"))

		if !fileExists(result.FFMpegPath) {

			if runtime.GOOS == "windows" {
				result.FFMpegPath = path.Join("C:\\ffmpeg\\bin\\", getBinaryFileName("ffmpeg"))
			} else {
				result.FFMpegPath = path.Join("/usr/bin", getBinaryFileName("ffmpeg"))
			}

			if !fileExists(result.FFMpegPath) {
				msg, _ := Localizer.Localize(&i18n.LocalizeConfig{
					DefaultMessage: &i18n.Message{
						ID:    "ErrorCodecBin",
						Other: "Error: Could not find the ffmpeg binary (ffmpeg)",
					},
				})
				fmt.Println(msg)
				msg, _ = Localizer.Localize(&i18n.LocalizeConfig{
					DefaultMessage: &i18n.Message{
						ID:    "FilesMissing",
						Other: "Seems like some required files are missing.",
					},
				})
				fmt.Println(msg)
				msg, _ = Localizer.Localize(&i18n.LocalizeConfig{
					DefaultMessage: &i18n.Message{
						ID:    "ReinstallFix",
						Other: "In order to fix this error, you could re-install PersonalMediaVault.",
					},
				})
				fmt.Println(msg)
				os.Exit(1)
			}

		}
	}

	if !fileExists(result.FFProbePath) {
		result.FFProbePath = path.Join(getDirName(), "bin", getBinaryFileName("ffprobe"))

		if !fileExists(result.FFProbePath) {

			if runtime.GOOS == "windows" {
				result.FFProbePath = path.Join("C:\\ffmpeg\\bin\\", getBinaryFileName("ffprobe"))
			} else {
				result.FFProbePath = path.Join("/usr/bin", getBinaryFileName("ffprobe"))
			}

			if !fileExists(result.FFProbePath) {
				msg, _ := Localizer.Localize(&i18n.LocalizeConfig{
					DefaultMessage: &i18n.Message{
						ID:    "ErrorProbeBin",
						Other: "Error: Could not find the ffprobe binary (ffprobe)",
					},
				})
				fmt.Println(msg)
				msg, _ = Localizer.Localize(&i18n.LocalizeConfig{
					DefaultMessage: &i18n.Message{
						ID:    "FilesMissing",
						Other: "Seems like some required files are missing.",
					},
				})
				fmt.Println(msg)
				msg, _ = Localizer.Localize(&i18n.LocalizeConfig{
					DefaultMessage: &i18n.Message{
						ID:    "ReinstallFix",
						Other: "In order to fix this error, you could re-install PersonalMediaVault.",
					},
				})
				fmt.Println(msg)
				os.Exit(1)
			}
		}
	}

	if result.H264Codec == "" {
		detectedCodec, err := detectH264Codec(result)

		if err == nil {
			result.H264Codec = detectedCodec
		} else {
			msg, _ := Localizer.Localize(&i18n.LocalizeConfig{
				DefaultMessage: &i18n.Message{
					ID:    "CodecWarning",
					Other: "Warning: Could not detect an encoder for H.264 in your FFMpeg installation. The will lead to errors when trying to encode videos.",
				},
			})
			fmt.Println(msg)

			result.H264Codec = H264_CODEC_DEFAULT
		}
	}

	return result
}

File: launcher-gui/src/utils/ffmpeg_config.rs
// FFmpeg config utils

use std::{
    collections::{HashMap, HashSet},
    env,
    fs::{self, create_dir_all, read_to_string},
    path::Path,
    process::Command,
};

use dirs::config_dir;

use crate::{
    log_debug,
    models::{
        FFmpegBadInstallationError, FFmpegConfig, VIDEO_CODEC_ALTERNATIVE, VIDEO_CODEC_DEFAULT,
    },
    utils::{file_exists, get_binary_name, get_dirname},
};

pub fn load_ffmpeg_config_from_file() -> Result<FFmpegConfig, ()> {
    let mut dir = match config_dir() {
        Some(d) => d,
        None => {
            return Err(());
        }
    };

    dir.push("PersonalMediaVault");

    if let Err(_) = create_dir_all(&dir) {
        return Err(());
    }

    dir.push("ffmpeg.json");

    let file_str = match read_to_string(dir) {
        Ok(s) => s,
        Err(_) => {
            return Err(());
        }
    };

    let config: FFmpegConfig = match serde_json::from_str(&file_str) {
        Ok(c) => c,
        Err(_) => {
            return Err(());
        }
    };

    Ok(config)
}

pub fn write_ffmpeg_to_config_file(config: &FFmpegConfig) -> Result<(), String> {
    let mut dir = match config_dir() {
        Some(d) => d,
        None => {
            return Err("Could not resolve the user configuration folder".to_string());
        }
    };

    dir.push("PersonalMediaVault");

    if let Err(e) = create_dir_all(&dir) {
        return Err(e.to_string());
    }

    dir.push("ffmpeg.json");

    let file_str = match serde_json::to_string(config) {
        Ok(s) => s,
        Err(e) => {
            return Err(e.to_string());
        }
    };

    if let Err(e) = fs::write(dir, file_str) {
        return Err(e.to_string());
    }

    Ok(())
}

fn detect_video_codecs(config: &FFmpegConfig) -> Result<HashSet<String>, String> {
    let mut cmd = Command::new(&config.ffmpeg_path);

    cmd.arg("-encoders");

    let out = match cmd.output() {
        Ok(o) => o,
        Err(e) => {
            return Err(e.to_string());
        }
    };

    let out_str = String::from_utf8_lossy(&out.stdout).to_string();

    let mut available_codecs: HashSet<String> = HashSet::new();

    let lines = out_str.split("\n");

    for line in lines {
        let line_parts: Vec<&str> = line.trim().split(" ").collect();

        if line_parts.len() < 2 {
            continue;
        }

        let codec_name = line_parts[1];

        available_codecs.insert(codec_name.to_string());
    }

    Ok(available_codecs)
}

fn detect_video_codec(config: &FFmpegConfig) -> Result<&'static str, String> {
    let available_codecs = detect_video_codecs(config)?;

    if available_codecs.contains(VIDEO_CODEC_DEFAULT) {
        return Ok(VIDEO_CODEC_DEFAULT);
    }

    if available_codecs.contains(VIDEO_CODEC_ALTERNATIVE) {
        return Ok(VIDEO_CODEC_ALTERNATIVE);
    }

    Err("Unavailable video codec".to_string())
}

/// Checks if a codec is in the list of available codecs
pub fn check_ffmpeg_codec(config: &FFmpegConfig) -> bool {
    match detect_video_codecs(config) {
        Ok(available_codecs) => available_codecs.contains(&config.video_codec),
        Err(_) => false,
    }
}

pub fn load_ffmpeg_config() -> Result<FFmpegConfig, FFmpegBadInstallationError> {
    let mut result = FFmpegConfig {
        ffmpeg_path: "".to_string(),
        ffprobe_path: "".to_string(),
        video_codec: "".to_string(),
    };

    if let Ok(c) = load_ffmpeg_config_from_file() {
        result = c;
    }

    if result.ffmpeg_path.is_empty() || !file_exists(&result.ffmpeg_path) {
        let mut dir = get_dirname();
        dir.push("bin");
        dir.push(get_binary_name("ffmpeg"));
        result.ffmpeg_path = dir.to_string_lossy().to_string();

        if !file_exists(&result.ffmpeg_path) {
            if env::consts::OS == "windows" {
                let p = Path::new("C:\\ffmpeg\\bin\\ffmpeg.exe");
                result.ffmpeg_path = p.to_string_lossy().to_string();
            } else {
                let p = Path::new("/usr/bin/ffmpeg");
                result.ffmpeg_path = p.to_string_lossy().to_string();
            }

            if !file_exists(&result.ffmpeg_path) {
                return Err(FFmpegBadInstallationError::FFmpegMissing);
            }
        }
    }

    if result.ffprobe_path.is_empty() || !file_exists(&result.ffprobe_path) {
        let mut dir = get_dirname();
        dir.push("bin");
        dir.push(get_binary_name("ffprobe"));
        result.ffprobe_path = dir.to_string_lossy().to_string();

        if !file_exists(&result.ffprobe_path) {
            if env::consts::OS == "windows" {
                let p = Path::new("C:\\ffmpeg\\bin\\ffprobe.exe");
                result.ffprobe_path = p.to_string_lossy().to_string();
            } else {
                let p = Path::new("/usr/bin/ffprobe");
                result.ffprobe_path = p.to_string_lossy().to_string();
            }

            if !file_exists(&result.ffprobe_path) {
                return Err(FFmpegBadInstallationError::FFprobeMissing);
            }
        }
    }

    if result.video_codec.is_empty() {
        result.video_codec = match detect_video_codec(&result) {
            Ok(c) => c.to_string(),
            Err(e) => {
                log_debug!("[WARNING] Could not detect a video encoder in your FFMpeg installation. The will lead to errors when trying to encode videos. DEtails: {e}");

                VIDEO_CODEC_DEFAULT.to_string()
            }
        }
    }

    Ok(result)
}
